File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionXor.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.FunctionContext;
import co.nlighten.jsontransform.functions.common.TransformerFunction;

public class TransformerFunctionXor extends TransformerFunction {
    public TransformerFunctionXor() {
        super();
    }
    @Override
    public Object apply(FunctionContext context) {
        var value = context.getJsonElementStreamer(null);
        var adapter = context.getAdapter();
        return value.stream().filter(adapter::isTruthy).count() == 1;
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionXor.ts
import TransformerFunction from "./common/TransformerFunction";
import FunctionContext from "./common/FunctionContext";
import { isTruthy } from "../JsonHelpers";

class TransformerFunctionXor extends TransformerFunction {
  constructor() {
    super({});
  }

  override async apply(context: FunctionContext): Promise<any> {
    const arr = await context.getJsonElementStreamer(null);
    if (arr == null) {
      return false;
    }
    return (
      (await arr
        .stream()
        .filter(item => isTruthy(item))
        .count()) === 1
    );
  }
}

export default TransformerFunctionXor;
